File: android/src/main/kotlin/io/adrop/adrop_ads/popupAd/FlutterAdropPopupAd.kt
package io.adrop.adrop_ads.popupAd

import android.app.Activity
import android.content.Context
import android.util.Log
import io.adrop.adrop_ads.AdType
import io.adrop.adrop_ads.AdropAd
import io.adrop.adrop_ads.bridge.AdropChannel
import io.adrop.adrop_ads.bridge.AdropMethod
import io.adrop.ads.model.AdropErrorCode
import io.adrop.ads.popupAd.AdropPopupAd
import io.adrop.ads.popupAd.AdropPopupAdListener
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel

class FlutterAdropPopupAd(
    context: Context,
    unitId: String,
    requestId: String,
    messenger: BinaryMessenger
) : AdropAd(), AdropPopupAdListener {

    private var popupAd: AdropPopupAd = AdropPopupAd(context, unitId)
    private var adropEventListenerChannel: MethodChannel?

    init {
        popupAd.popupAdListener = this
        val channelName = AdropChannel.adropEventListenerChannelOf(AdType.Popup, requestId)
        adropEventListenerChannel = if (channelName != null) {
            MethodChannel(messenger, channelName)
        } else {
            null
        }
    }

    override fun load() {
        popupAd.load()
    }

    override fun show(activity: Activity) {
        popupAd.show(activity)
    }

    fun customize(data: Map<String, Any>) {
        data["closeTextColor"]?.let { popupAd.closeTextColor = colorOf(it) }
        data["hideForTodayTextColor"]?.let { popupAd.hideForTodayTextColor = colorOf(it) }
        data["backgroundColor"]?.let { popupAd.backgroundColor = colorOf(it) }
    }

    private fun colorOf(color: Any): Int {
        return when (color) {
            is Long -> color.toInt()
            is Int -> color
            else -> 0
        }
    }

    override fun destroy() {
        popupAd.destroy()
    }

    override fun onAdFailedToReceive(ad: AdropPopupAd, errorCode: AdropErrorCode) {
        adropEventListenerChannel?.invokeMethod(
            AdropMethod.DID_FAIL_TO_RECEIVE_AD,
            mapOf("errorCode" to errorCode.name)
        )
    }

    override fun onAdReceived(ad: AdropPopupAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_RECEIVE_AD, null)
    }

    override fun onAdClicked(ad: AdropPopupAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_CLICK_AD, null)
    }

    override fun onAdImpression(ad: AdropPopupAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_IMPRESSION, null)
    }

    override fun onAdFailedToShowFullScreen(ad: AdropPopupAd, errorCode: AdropErrorCode) {
        adropEventListenerChannel?.invokeMethod(
            AdropMethod.DID_FAIL_TO_SHOW_FULL_SCREEN,
            mapOf("errorCode" to errorCode.name)
        )
    }

    override fun onAdDidPresentFullScreen(ad: AdropPopupAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_PRESENT_FULL_SCREEN, null)
    }

    override fun onAdDidDismissFullScreen(ad: AdropPopupAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_DISMISS_FULL_SCREEN, null)
    }
}

File: ios/Classes/popupAd/FlutterAdropPopupAd.swift
import Foundation
import Flutter
import AdropAds

class FlutterAdropPopupAd: NSObject, AdropAd, AdropPopupAdDelegate {
    
    private let messenger:FlutterBinaryMessenger
    private let adropEventListenerChannel: FlutterMethodChannel?
    private let requestId: String
    private let popupAd: AdropPopupAd
    
    init(
        unitId: String,
        requesetId: String,
        messenger: FlutterBinaryMessenger) {
            self.messenger = messenger
            self.requestId = requesetId
            let methodChannelName = AdropChannel.adropEventListenerChannel(adType: AdType.popup, id: requesetId)
            self.adropEventListenerChannel = methodChannelName != nil ? FlutterMethodChannel(name: methodChannelName!, binaryMessenger: messenger) : nil
            
            self.popupAd = AdropPopupAd(unitId: unitId)
        }
    
    func load() {
        self.popupAd.delegate = self
        popupAd.load()
    }
    
    func show() {
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene else {
            return
        }
        
        guard let rootViewController = windowScene.windows.first?.rootViewController else {
            return
        }
        
        popupAd.show(fromRootViewController: rootViewController)
    }
    
    func customize(_ data: [String:Any]) {
        if let closeTextColor = data["closeTextColor"] as? UInt32 {
            popupAd.closeTextColor = UIColor(fromFlutterColor: closeTextColor)
        }
        if let hideForTodayTextColor = data["hideForTodayTextColor"] as? UInt32 {
            popupAd.hideForTodayTextColor = UIColor(fromFlutterColor: hideForTodayTextColor)
        }
        if let backgroundColor = data["backgroundColor"] as? UInt32 {
            popupAd.backgroundColor = UIColor(fromFlutterColor: backgroundColor)
        }
    }
    
    func onAdReceived(_ ad: AdropPopupAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_RECEIVE_AD, arguments: nil)
    }
    
    func onAdFailedToReceive(_ ad: AdropPopupAd, _ errorCode: AdropErrorCode) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_FAILED_TO_RECEIVE, arguments: ["errorCode":AdropErrorCodeToString(code: errorCode)])
    }
    
    
    func onAdImpression(_ ad: AdropPopupAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_IMPRESSION, arguments: nil)
    }
    
    func onAdClicked(_ ad: AdropPopupAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_CLICK_AD, arguments: nil)
    }
    
    func onAdWillPresentFullScreen(_ ad: AdropPopupAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.WILL_PRESENT_FULL_SCREEN, arguments: nil)
    }
    
    func onAdDidPresentFullScreen(_ ad: AdropPopupAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_PRESENT_FULL_SCREEN, arguments: nil)
    }
    
    func onAdWillDismissFullScreen(_ ad: AdropPopupAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.WILL_DISMISS_FULL_SCREEN, arguments: nil)
    }
    
    func onAdDidDismissFullScreen(_ ad: AdropPopupAd) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_DISMISS_FULL_SCREEN, arguments: nil)
    }
    
    func onAdFailedToShowFullScreen(_ ad: AdropPopupAd, _ errorCode: AdropErrorCode) {
        adropEventListenerChannel?.invokeMethod(AdropMethod.DID_FAIL_TO_SHOW_FULL_SCREEN, arguments: ["errorCode": AdropErrorCodeToString(code: errorCode)])
    }
}
